### PR TITLE
Improvement : constant pool now less concurrent

### PR DESCRIPTION
**Pull request by doom369**
Motivation : 

Current constant pool holds all data within `HashMap` and all access to this `HashMap` is done via `synchronized` blocks. Thus `CuncurrentHashMap` will be here more efficient as it designed for higher throughput and will use less locks. Also `valueOf` method was not very efficient as it performed `get` operation 2 times.

Modifications : 

`HashMap -> PlatformDependent.newConcurrentHashMap()`.
`ValueOf` is more efficient now, threadsafe and uses less locks. Downside is that `final T tempConstant = newConstant(nextId(), name);` could be called more than 1 time during high contention. 

Result : 

Less contention, cleaner code.

P. S. Just found also that `public T newInstance(String name)` method wasn't threadsafe. However in this PR this is not fixed.

### CODE REVIEW

File: common/src/main/java/io/netty/util/ConstantPool.java [hunk 1]
```diff
@@ -55,52 +56,53 @@ public T valueOf(Class<?> firstNameComponent, String secondNameComponent) {
...
-                c = newInstance0(name);
+    /**
+     * Get existing constant by name or creates new one if not exists. Threadsafe
+     *
+     * @param name the name of the {@link Constant}
+     */
+    private T getOrCreate(String name) {
+        T constant = constants.get(name);
+        if (constant == null) {
+            final T tempConstant = newConstant(nextId(), name);
```
Comment (johnou) [inline common/src/main/java/io/netty/util/ConstantPool.java]: Does it matter if we skip (discard) ids?

Comment (doom369) [inline common/src/main/java/io/netty/util/ConstantPool.java]: @johnou maybe. I haven't yet dig enough to say for sure. I don't to bring any regression so changing only things I'm 100% sure.

File: common/src/main/java/io/netty/util/ConstantPool.java [hunk 2]
```diff
@@ -55,52 +56,53 @@ public T valueOf(Class<?> firstNameComponent, String secondNameComponent) {
...
+        return constants.containsKey(name);
     }
 
     /**
      * Creates a new {@link Constant} for the given {@code name} or fail with an
      * {@link IllegalArgumentException} if a {@link Constant} for the given {@code name} exists.
      */
-    @SuppressWarnings("unchecked")
     public T newInstance(String name) {
         if (exists(name)) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ConstantPool.java]: It looks like there is a preexisting bug here.  We don't hold the lock throughout this operation and so the end result may be undefined.

Thread1 -> exists(name) == false
Thread2 -> exists(name) == false
Thread1 -> put(name, value1)
Thread2 -> put(name, value2)

At this point is it undefined what value will be associated with `name` ... which is looks like the intention here is to provide more deterministic behavior. We should assume the "correct"/"deterministic" behavior exists before this PR and preserve this behavior (one should fail, and one should succeeded)

Comment (doom369) [inline common/src/main/java/io/netty/util/ConstantPool.java]: @Scottmitch correct. I pointed this problem in PR description. I think it should be fixed with separate PR.

Comment (doom369) [inline common/src/main/java/io/netty/util/ConstantPool.java]: @Scottmitch I can fix it.

Comment (doom369) [inline common/src/main/java/io/netty/util/ConstantPool.java]: @Scottmitch with ConcurrentMap there is no issue anymore.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ConstantPool.java]: @doom369 - I missed that in the PR description :+1:. I think we should fix this first and then consider "improvements". So for this PR that is why I recommended assume it is implemented correctly and preserve this behavior (just to be sure we can preserve the behavior with the new approach).

**Comment by doom369** [comment]
@Scottmitch fixed. please review. I used same approach as for `valueOf`

**Comment by trustin** [comment]
`ConstantPool` is not meant to be called frequently. Most likely only at class-loading time. Once they are instantiated, they are meant to be accessed via direct reference (e.g. `MyAttributeKeys.MY_KEY`)

If there's a use case we didn't think of, I'd be curious to know.

**Comment by doom369** [comment]
Ah... I see. I thought `ConstantPool` is related somehow with `channel().attr()` as it operates with `AttributeKey`. Now it see there one more storage for constants within channel. Ok.

But anyway I think ConcurrentMap is faster better here :).

**Comment by johnou** [comment]
@doom369 how many constant pools are there?

**Comment by doom369** [comment]
@johnou 3

In AttributeKey, Signal, ChannelOption

File: common/src/main/java/io/netty/util/ConstantPool.java [hunk 3]
```diff
@@ -115,12 +125,8 @@ private String checkNotNullAndNotEmpty(String name) {
 
     protected abstract T newConstant(int id, String name);
 
-    @Deprecated
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ConstantPool.java]: can we leave this as deprecated?

File: common/src/main/java/io/netty/util/ConstantPool.java [hunk 4]
```diff
@@ -115,12 +125,8 @@ private String checkNotNullAndNotEmpty(String name) {
...
     public final int nextId() {
-        synchronized (constants) {
-            int id = nextId;
-            nextId++;
-            return id;
-        }
+        return nextId.getAndIncrement();
     }
+
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ConstantPool.java]: nit: kill this line

**Comment by Scottmitch** [comment]
few more comments and lgtm

**Comment by doom369** [comment]
@Scottmitch done

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (c6a13e28b93b7b4d1f99f556cc894765a4c8e2f1)
